Update task list items to show spinner on selection

yashitamittal11 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

### Summary:
This pull request introduces a spinner component to indicate the loading status when the user clicks the product task.

### Reason for the change:
When clicking on the product tasks, there is a noticeable delay before being taken to the next page.
Previously, users were left waiting without any indication of the loading status.

### Steps to fix:

This PR introduces a spinner component to indicate the loading status during the async actions on clicking the product task. Although the loading time optimization is not included in this PR, the spinner serves as an initial improvement to provide visual feedback to users during the loading process.

Original behaviour - No Spinner shows on task selection

![No Spinner before](https://github.com/woocommerce/woocommerce/assets/21078987/776f202c-ac6d-465f-8538-6684cb3d5177)

After fix - Spinner displays showing loading status on selection

![Spinner shows on selection](https://github.com/woocommerce/woocommerce/assets/21078987/7f44ac0d-65e4-4c7e-a683-69ec6d3ed61b)


Closes https://github.com/woocommerce/woocommerce/issues/36510

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Case 1: Show spinner

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->


1. Pull down this branch on a new or existing site.
2. Go to WooCommerce > Home and click on `Add Products` to open the product task list.
3. Click on any product task (For example- Physical product) 
4. Ensure the spinner shows until the page redirects to another page. 
5. If the redirection is unsuccessful due to an error, the spinner is hidden.

https://github.com/woocommerce/woocommerce/assets/21078987/98ce0770-dcda-485e-af6a-93ec285f7fbf

Case 2: Hide spinner

If the page redirection fails on clicking the product task list item, we should hide the spinner so it doesn't show indefinitely. To test this case, I changed the variable's value to null while debugging the app to generate the error.

1. First, set up WooCommerce in the dev mode. Clone the repo from [this](https://github.com/woocommerce/woocommerce) link.
2. Run `nvm use` and `pnpm install` from plugins/woocommerce
3. Start the environment by running `pnpm -- wp-env start`
4. Go to plugins/woocommerce-admin and run `pnpm dev` to build the dev version, as we need to debug the code in the WooCommerce-admin.
5. Once the page loads, follow the steps in the video below to generate the error and ensure spinner is hidden.


https://github.com/woocommerce/woocommerce/assets/21078987/e28f4d72-1181-42f0-99bb-692a91dbc961



<!-- End testing instructions -->

Commit:
Update task list to show a spinner on item click

There is an issue of a noticeable delay when clicking on product task list items, which leaves users waiting without any indication of the loading status.
To enhance the user experience, add a loading spinner that provides visual feedback during the transition to the next page.

Commit:
Remove spinner from product task list on error

Implemented logic to remove the loading spinner if an error occurs on clicking product task list, preventing it from being displayed indefinitely.

Commit:
Fix linting errors in the CSS

timur987 commented on plugins/woocommerce-admin/client/task-lists/fills/products/index.tsx:
```diff
@@ -52,6 +53,11 @@ const ViewControlButton: React.FC< {
 	const [ isExpanded, setIsExpanded ] = useState< boolean >( false );
+
+	// State to track whether the task list item is clicked and show a spinner
+	const [ isTaskListItemclicked, setIsTaskListItemclicked ] =
```
I can be wrong, but I think that it's more important to highlight that a spinner is shown rather than that a list item is clicked. What do you think?

yashitamittal11 replied:
I agree. After reviewing the definitions of other state variables in the code, it appears that the variable names are self-explanatory without the need for preceding comments. Therefore, I think it would make sense to remove the comment in question. It would ensure consistency throughout the codebase. What do you think?

timur987 replied:
I thought about renaming the variable to highlight it's needed to show a spinner rather that reflect the state of a clicked button.

timur987 commented on plugins/woocommerce-admin/client/task-lists/fills/products/use-product-types-list-items.tsx:
```diff
@@ -15,11 +15,15 @@ const useProductTypeListItems = (
+		setIsTaskListItemclicked,
 	}: {
 		onClick?: () => void;
+		setIsTaskListItemclicked?: ( value: boolean ) => void;
```
I wonder do we need to introduce yet another onClick function? What about adding your `setIsTaskListItemclicked` call inside already existing onClick calls like converting
```js
onClick: someFunction
```

into
```js
onclick: {
    someFunction();
    setIsTaskListItemclicked( true );
}
```

In that case you will use the same entry point for handling clicks and also won't need to replace missing function with `() => undefined`?

yashitamittal11 replied:
Previously, an additional isTaskListItemClicked variable was introduced to handle the logic for showing the spinner.
However, it mimics the existing isRequesting state variable. Since the logic is now changed, I am resolving this conversation.

timur987 commented on plugins/woocommerce-admin/client/task-lists/fills/products/stack.scss:
```diff
@@ -79,4 +80,22 @@
+		left: 0;
+		right: 0;
+		bottom: 0;
+		background-color: rgba(255, 255, 255, 0.8);
```
We might probably need to consult our designers regarding to styling.

yashitamittal11 replied:
That sounds right. Also, I checked the code and noticed we are using HEX codes. So I replaced the background color with this code -
`background-color: #fff;
 opacity: 0.8;`

timur987 replied:
On the other hand, since we've already used the same value for background [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/task-lists/fills/components/load-sample-product-modal.scss#L3), it might be good enough to go ahead with this overlay color.

Commit:
Refactor code to use isRequesting variable for spinner

Previously, an additional isTaskListItemClicked variable was introduced to handle the logic for showing spinner.
However, it mimics the existing isRequesting state variable. To simplify the code and avoid using two variables simultaneously,
this commit updates the logic to utilize the existing isRequesting variable.

Commit:
Add changelog file

timur987 commented on plugins/woocommerce-admin/client/task-lists/fills/products/stack.tsx:
```diff
@@ -21,17 +21,24 @@ type StackProps = {
 
 	return (
 		<div className="woocommerce-products-stack">
+			{ isTaskListItemclicked && (
```
The last piece of feedback (I hope so) is to fix camel case there and rename it from `isTaskListItemclicked` to `isTaskListItemClicked` wherever it's used.

yashitamittal11 replied:
Thanks for pointing it out. I have fixed it.

Commit:
Fix camelCase in isTaskListItemclicked variable

Commit:
Merge remote-tracking branch 'origin/trunk' into update/36510-task-list-items

Commit:
Add return statement to maintain spinner state on redirection

- Add the return statement to prevent the update of the state variable responsible for showing the spinner during redirection.
- Previously, the state variable was being reset, causing the spinner to disappear before redirecting to the product page.

rjchow reviewed (approved):
Looks good to me! Thanks @timur987 for the thorough review before, too!

yashitamittal11 commented:
Thanks for the review. 👍
@rjchow I am not authorized to merge this pull request. Can you please merge it? 🙏

rjchow commented:
> Thanks for the review. 👍 @rjchow I am not authorized to merge this pull request. Can you please merge it? 🙏

Done, thanks for your work on this!
